crates/chia-datalayer/src/merkle.rs

```diff
@@ -0,0 +1,1997 @@
+#[cfg(feature = "py-bindings")]
+use pyo3::{
+    buffer::PyBuffer,
+    exceptions::{PyAttributeError, PyValueError},
+    pyclass, pymethods, PyResult, Python,
+};
+
+use clvmr::sha2::Sha256;
+use num_traits::ToBytes;
+use std::cmp::Ordering;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::iter::zip;
+use std::mem::size_of;
+use std::ops::Range;
+
+type TreeIndex = u32;
+type Parent = Option<TreeIndex>;
+type Hash = [u8; 32];
+type KvId = i64;
+
+// assumptions
+// - root is at index 0
+// - any case with no keys will have a zero length blob
+
+const fn range_by_length(start: usize, length: usize) -> Range<usize> {
+    start..start + length
+}
+const fn max(left: usize, right: usize) -> usize {
+    [left, right][(left < right) as usize]
+}
+// TODO: once not experimental...  something closer to this
+// const fn max<T: ~const std::cmp::PartialOrd>(left: T, right: T) -> T { if left < right {right} else {left} }
+
+// TODO: consider in more detail other serialization tools such as serde and streamable
+// define the serialized block format
+const METADATA_RANGE: Range<usize> = 0..METADATA_SIZE;
+const TYPE_RANGE: Range<usize> = range_by_length(0, size_of::<u8>());
+const DIRTY_RANGE: Range<usize> = range_by_length(TYPE_RANGE.end, size_of::<u8>());
+const METADATA_SIZE: usize = DIRTY_RANGE.end;
+
+// common fields
+const HASH_RANGE: Range<usize> = range_by_length(0, size_of::<Hash>());
+// const PARENT_RANGE: Range<usize> = range_by_length(HASH_RANGE.end, size_of::<TreeIndex>());
+const PARENT_RANGE: Range<usize> = range_by_length(HASH_RANGE.end, size_of::<TreeIndex>());
+// internal specific fields
+const LEFT_RANGE: Range<usize> = range_by_length(PARENT_RANGE.end, size_of::<TreeIndex>());
+const RIGHT_RANGE: Range<usize> = range_by_length(LEFT_RANGE.end, size_of::<TreeIndex>());
+// leaf specific fields
+const KEY_RANGE: Range<usize> = range_by_length(PARENT_RANGE.end, size_of::<KvId>());
+const VALUE_RANGE: Range<usize> = range_by_length(KEY_RANGE.end, size_of::<KvId>());
+
+const DATA_SIZE: usize = max(RIGHT_RANGE.end, VALUE_RANGE.end);
+const BLOCK_SIZE: usize = METADATA_SIZE + DATA_SIZE;
+type BlockBytes = [u8; BLOCK_SIZE];
+type MetadataBytes = [u8; METADATA_SIZE];
+type DataBytes = [u8; DATA_SIZE];
+const DATA_RANGE: Range<usize> = METADATA_SIZE..METADATA_SIZE + DATA_SIZE;
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+#[repr(u8)]
+pub enum NodeType {
+    Internal = 0,
+    Leaf = 1,
+}
+
+impl NodeType {
+    pub fn from_u8(value: u8) -> Result<Self, String> {
+        match value {
+            // ha!  feel free to laugh at this
+            x if (NodeType::Internal as u8 == x) => Ok(NodeType::Internal),
```

Comment:
I really think this should either be `0 => Ok(NodeType::Internal)` and so on, or use a macro to generate both conversions. Same with `to_u8`.

Reply:
macro in [`825f7ac` (#673)](https://github.com/Chia-Network/chia_rs/pull/673/commits/825f7ac3c3ec2af506dbb9559ef9f380c924c053)